### PR TITLE
CB-16217 Validate environment edit when the user try to remove a used…

### PR DESCRIPTION
**Pull request by topolyai5**
… subnet id

See detailed description in the commit message.

### CODE REVIEW

File: environment/src/test/java/com/sequenceiq/environment/environment/validation/validators/SubnetUsageValidatorTest.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        when(stackV4Endpoint.getUsedSubnetsByEnvironment(0L, ENVIRONMENT_CRN))
+                .thenReturn(new UsedSubnetsByEnvironmentResponse(List.of(createUsedSubnetWithCrn("sub1", "WORKLOAD"))));
+        when(freeIpaV1Endpoint.getUsedSubnetsByEnvironment(ENVIRONMENT_CRN))
+                .thenReturn(new UsedSubnetsByEnvironmentResponse(List.of(createUsedSubnetWithCrn("sub2", "FREEIPA"))));
+        when(databaseServerV4Endpoint.getUsedSubnetsByEnvironment(ENVIRONMENT_CRN))
+                .thenReturn(new UsedSubnetsByEnvironmentResponse(List.of(createUsedSubnetWithCrn("sub3", "REDBEAMS"))));
+        underTest.validate(environment, network, resultBuilder);
+        ValidationResult validationResult = resultBuilder.build();
+        Assertions.assertTrue(validationResult.hasError());
+        Assertions.assertEquals("1. The Data Hub uses the subnets of [sub1], cannot remove these from the environment\n" +
```
Comment (doktoric) [inline environment/src/test/java/com/sequenceiq/environment/environment/validation/validators/SubnetUsageValidatorTest.java:87]: Can we add the datalake as well ?

**Review by doktoric — CHANGES_REQUESTED** [review]
added one ask

File: core/src/main/java/com/sequenceiq/cloudbreak/repository/InstanceMetaDataRepository.java [hunk 2]
```diff
@@ -332,4 +333,11 @@ List<InstanceMetadataViewDelegate> findInstanceMetadataViewsByStackIdAndPrivateI
             "WHERE s.id = :stackId " +
             "ORDER BY im.privateId desc")
     Page<Long> findLastPrivateIdForStack(@Param("stackId") Long stackId, Pageable page);
+
+    @Query("SELECT distinct new com.sequenceiq.cloudbreak.dto.SubnetIdWithResourceNameAndCrn(s.name, s.resourceCrn, im.subnetId, s.type) " +
+            "FROM InstanceMetaData im " +
+            "LEFT JOIN im.instanceGroup ig  " +
+            "LEFT JOIN ig.stack s " +
+            "WHERE s.environmentCrn = :environmentCrn")
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/repository/InstanceMetaDataRepository.java:341]: wouldn't this return deleted/terminated instances/stacks?

File: freeipa/src/main/java/com/sequenceiq/freeipa/repository/InstanceMetaDataRepository.java [hunk 3]
```diff
@@ -49,4 +50,11 @@ public interface InstanceMetaDataRepository extends CrudRepository<InstanceMetaD
 
     List<InstanceMetaData> findAllByInstanceGroupAndInstanceStatus(InstanceGroup instanceGroup, InstanceStatus status);
 
+    @Query("SELECT distinct new com.sequenceiq.common.model.SubnetIdWithResourceNameAndCrn(s.name, s.resourceCrn, im.subnetId, 'FREEIPA') " +
+            "FROM InstanceMetaData im " +
+            "LEFT JOIN im.instanceGroup ig  " +
+            "LEFT JOIN ig.stack s " +
+            "WHERE s.environmentCrn = :environmentCrn")
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/repository/InstanceMetaDataRepository.java:57]: same here, I don't see any filtering for deleted instances

File: common-model/src/main/java/com/sequenceiq/common/api/UsedSubnetWithResourceResponse.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
+package com.sequenceiq.common.api;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class UsedSubnetWithResourceResponse {
```
Comment (lacikaaa) [inline common-model/src/main/java/com/sequenceiq/common/api/UsedSubnetWithResourceResponse.java:6]: could you add toString impl to the model classes?

File: core/src/main/java/com/sequenceiq/cloudbreak/repository/InstanceMetaDataRepository.java (hunk 2 above)
Comment (topolyai5) [reply core/src/main/java/com/sequenceiq/cloudbreak/repository/InstanceMetaDataRepository.java:341]: good catch, thanks

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/stacks/StackV4Endpoint.java [hunk 5]
```diff
@@ -676,4 +678,13 @@ FlowIdentifier verticalScalingByName(
...
             @Valid StackVerticalScaleV4Request updateRequest);
+
+    @GET
+    @Path("internal/used_subnets")
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(value = GET_USED_SUBNETS_BY_ENVIRONMENT_CRN, produces = MediaType.APPLICATION_JSON, notes = Notes.STACK_NOTES,
+            nickname = "getUsedSubnetsByEnvironment")
+    UsedSubnetsByEnvironmentResponse getUsedSubnetsByEnvironment(
+            @PathParam("workspaceId") Long workspaceId,
+            @QueryParam("environmentCrn") String environmentCrn);
```
Comment (lacikaaa) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/stacks/StackV4Endpoint.java:689]: could you add `ValidCrn` validation to these endpoints environmentcrn parameter?

File: common-model/src/main/java/com/sequenceiq/common/api/UsedSubnetsByEnvironmentResponse.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package com.sequenceiq.common.api;
+
+import java.util.List;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class UsedSubnetsByEnvironmentResponse {
```
Comment (lacikaaa) [inline common-model/src/main/java/com/sequenceiq/common/api/UsedSubnetsByEnvironmentResponse.java:8]: please don't forget tostring for this one too

**Review by lacikaaa — APPROVED** [review]

**Review by lturcsanyi — APPROVED** [review]

**Review by doktoric — APPROVED** [review]

**Comment by lnardai** [comment]
Please resolve conflicts.

**Pull request merged**